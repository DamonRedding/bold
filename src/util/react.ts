import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
