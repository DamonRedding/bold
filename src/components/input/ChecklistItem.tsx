import * as React from 'react'
import { CheckboxProps, Checkbox } from './Checkbox/Checkbox'
import { List } from 'immutable'

export interface ChecklistItemProps extends CheckboxProps {
    optionValue: any
}

export class ChecklistItem extends React.Component<ChecklistItemProps> {

    getCurrentValues = () => {
        const { value } = this.props

        let previousValues = []

        if (value !== undefined && value !== '') {
            previousValues = value
        }

        const currentValues = Array.isArray(previousValues) ? [...previousValues] : List.isList(previousValues) ? previousValues : [previousValues]

        return currentValues
    }

    handleChange = (event) => {
        const { onChange } = this.props
        let values: any = this.getCurrentValues()

        if (event.target.checked) {
            if (List.isList(values)) {
                values = values.push(this.props.optionValue)
            } else {
                values.push(this.props.optionValue)
            }
        } else {
            if (List.isList(values)) {
                values = values.splice(values.indexOf(this.props.optionValue), 1)
            } else {
                values.splice(values.indexOf(this.props.optionValue), 1)
            }
        }

        return onChange(values as any)
    }

    render() {
        const { optionValue, ...rest } = this.props
        const values = this.getCurrentValues()

        const isChecked = values.indexOf(optionValue) > -1
        return (
            <Checkbox
                {...rest}
                onChange={this.handleChange}
                checked={isChecked}
                onBlur={this.onBlur}
                value={optionValue}
            />
        )
    }

    private onBlur = () => { }
}
