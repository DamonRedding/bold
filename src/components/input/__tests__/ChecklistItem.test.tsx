import * as React from 'react'
import { shallowRenderAndMatch } from '../../../__tests__/test.defaults'
import { ChecklistItem } from '../ChecklistItem'
import { List } from 'immutable'
import { shallow } from 'enzyme'

describe('ChecklistItem', () => {
    it('marcado utilizando array', function () {
        shallowRenderAndMatch(<ChecklistItem label='check' optionValue='1' value={['1']} />)
    })
    it('marcado utilizando Immutable.List', function () {
        shallowRenderAndMatch(<ChecklistItem label='check' optionValue='1' value={List(['1'])} />)
    })
    it('deve conter a classe css "disabled"', function () {
        shallowRenderAndMatch(<ChecklistItem label='check' optionValue='1' disabled />)
    })
    it('deve renderizar corretamente', function () {
        shallowRenderAndMatch(<ChecklistItem label='check' optionValue='VALOR' />)
    })
    it('deve conter a classe css "disabled"', function () {
        shallowRenderAndMatch(<ChecklistItem label='check' disabled optionValue='VALOR' />)
    })
    it('deve adicionar o valor Ã  lista', function () {
        const onChange = jest.fn()

        const wrapper = shallow(<ChecklistItem label='check' optionValue='VALOR' onChange={onChange} />)

        wrapper.find('Checkbox').simulate('change', { target: { checked: true } })

        expect(onChange).toBeCalledWith(['VALOR'])
    })
    it('deve remover o valor da lista', function () {
        const onChange = jest.fn()

        const wrapper = shallow(<ChecklistItem value={['VALOR']} label='check' optionValue='VALOR' onChange={onChange} />)

        wrapper.find('Checkbox').simulate('change', { target: { checked: false } })

        expect(onChange).toBeCalledWith([])
    })
})
