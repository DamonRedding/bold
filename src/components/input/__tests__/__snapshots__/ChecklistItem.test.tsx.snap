// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChecklistItem deve conter a classe css "disabled" 1`] = `
<Checkbox
  checked={false}
  disabled={true}
  label="check"
  onBlur={[Function]}
  onChange={[Function]}
  value="1"
/>
`;

exports[`ChecklistItem deve conter a classe css "disabled" 2`] = `
<Checkbox
  checked={false}
  disabled={true}
  label="check"
  onBlur={[Function]}
  onChange={[Function]}
  value="VALOR"
/>
`;

exports[`ChecklistItem deve renderizar corretamente 1`] = `
<Checkbox
  checked={false}
  label="check"
  onBlur={[Function]}
  onChange={[Function]}
  value="VALOR"
/>
`;

exports[`ChecklistItem marcado utilizando Immutable.List 1`] = `
<Checkbox
  checked={true}
  label="check"
  onBlur={[Function]}
  onChange={[Function]}
  value="1"
/>
`;

exports[`ChecklistItem marcado utilizando array 1`] = `
<Checkbox
  checked={true}
  label="check"
  onBlur={[Function]}
  onChange={[Function]}
  value="1"
/>
`;
